src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs

```diff
@@ -1,93 +1,60 @@
-                    {
-                        IsAvailable = isAvailableProperty != null && (bool)isAvailableProperty.GetValue(null, new object[] { });
-                    }
-                    catch (TargetInvocationException)
-                    {
-                        IsAvailable = false;
-                    }
-                }
-            }
+            return TryGetField(() => RoleEnvironment.GetConfigurationSettingValue(configurationSettingName), out value);
         }
 
-        /// <summary>
-        /// Delegate the call because we don't want RoleEnvironment appearing in the function scope of the caller because that
-        /// would trigger the assembly load: the very thing we are attempting to avoid
-        /// </summary>
-        /// <param name="configurationSettingName"></param>
-        /// <returns></returns>
-        public static string GetConfigurationSettingValue(string configurationSettingName)
+        public static bool TryGetDeploymentId(out string id)
         {
-            return RoleEnvironment.GetConfigurationSettingValue(configurationSettingName);
+            return TryGetField(() => RoleEnvironment.DeploymentId, out id);
         }
 
-        public static string GetDeploymentId()
+        public static bool TryGetLocalResourceRootPath(string name, out string path)
         {
-            if (IsAvailable)
-            {
-                return RoleEnvironment.DeploymentId;
-            }
-
-            return string.Empty;
+            return TryGetField(() => RoleEnvironment.GetLocalResource(name).RootPath, out path);
         }
 
-        public static string GetLocalResourceRootPath(string name)
+        private static bool TryGetField<T>(Func<T> getValue, out T value)
         {
-            return RoleEnvironment.GetLocalResource(name).RootPath;
-        }
-
-        /// <summary>
-        /// Loads and returns the latest available version of the service runtime assembly.
-        /// </summary>
-        /// <returns>Loaded assembly, if any.</returns>
-        private static Assembly GetServiceRuntimeAssembly()
-        {
-            Assembly assembly = null;
+            value = default(T);
 
             try
             {
-                assembly = Assembly.LoadWithPartialName(_serviceRuntimeAssembly);
+                if (!RoleEnvironment.IsAvailable)
+                {
+                    // If RoleEnvironment isn't available, we can't access it.
+                    return false;
+                }
+
+                value = getValue();
+                return true;
             }
             catch (Exception e)
             {
+                // If the assembly is not available, a file load exception will be thrown.
+                // Otherwise, it must be an exception thrown by the loaded assembly itself.
                 if (!(e is FileNotFoundException || e is FileLoadException || e is BadImageFormatException))
                 {
                     throw;
                 }
```

Comment:
If these methods are called often, this catching will be expensive. We should try to call once, and if we get one of these exceptions, we should enable a circuit breaker.

Reply:
added...deploying to DEV again to test